quint/src/effects/modeChecker.ts

```diff
@@ -109,18 +109,34 @@ function modeConstraint(mode: OpQualifier): string {
   }
 }
 
+const componentKindPriority: ComponentKind[] = ['temporal', 'update', 'read']
+const componentDescription = new Map<ComponentKind, string>([
+  ['temporal', 'performs temporal operations over'],
+  ['update', 'updates'],
+  ['read', 'reads'],
+])
+const modesForArrow = new Map<ComponentKind, OpQualifier>([
+  ['temporal', 'temporal'],
+  ['update', 'action'],
+  ['read', 'def'],
+])
+const modesForConcrete = new Map<ComponentKind, OpQualifier>([
+  ['temporal', 'temporal'],
+  ['update', 'action'],
+  ['read', 'val'],
```

Comment:
why does 'read' correspond to `def` in one case, and to `val` in another? What happens if pass another operator as an argument?

Reply:
Arrow effects will have `def` mode, since the operator takes arguments, while a concrete effect will have `val`. I don't see how passing another operator as an argument would affect that, can you elaborate?

Reply:
Maybe I am mixing things up with modes. What happens if we have a `def` that accepts an operator as an argument and that operator happens to be `temporal`?

Reply:
If the `def` actually uses that operator in its body, it would produce a temporal effect in its body and an error would be reported saying `def` should be `temporal`.

This is what I believe should happen, I'll test it later to make sure this is working correctly.

Reply:
Ok, I'm testing this and actually there's no way to strict the mode of an operator given as argument since we don't allow effect signatures. This is what I wrote:

```scala
  def applyTemporal(f: (int) => bool): bool = {
    f(1)
  }

  temporal a = applyTemporal(x => always(x > 1))
```

There's no way to tell whether `f` is temporal or not, and there is no way to strict it as well.

This is bad, as we can also use this to apply action operations:

```ocaml
  action b = applyTemporal(x => z' = 1)
```

Perhaps we should enforce that no high order operators can have those modes in their arguments? It's not clear for me what consequences could arise from that, it might prevent users from doing something that they should be able to. We can try!

Either way, is it ok if I open a separate issue for this and we move on here? This is not a change introduced by this refactoring.